File: 63.unique-paths-ii.py
#
# @lc app=leetcode id=63 lang=python3
#
# [63] Unique Paths II
#

# @lc code=start
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        
        row_size = len(obstacleGrid)
        col_size = len(obstacleGrid[0])

        paths_matrix = [[0 for _ in range(col_size)] for _ in range(row_size)]

        if obstacleGrid[0][0] == 1:
            return 0

        paths_matrix[0][0] = 1

        for row in range(1, row_size):
            if obstacleGrid[row][0] == 0:
                paths_matrix[row][0] = paths_matrix[row - 1][0]

        for col in range(1, col_size):
            if obstacleGrid[0][col] == 0:
                paths_matrix[0][col] = paths_matrix[0][col - 1]

        for row in range(1, row_size):
            for col in range(1, col_size):
                if obstacleGrid[row][col] == 0:
                    paths_matrix[row][col] = paths_matrix[row - 1][col] + paths_matrix[row][col - 1]

        return paths_matrix[row_size - 1][col_size - 1]


        
# @lc code=end

File: 63.unique-paths-ii.java
/*
 * @lc app=leetcode id=63 lang=java
 *
 * [63] Unique Paths II
 */

// @lc code=start
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid[0][0] == 1) {
            return 0;
        }

        int rowSize = obstacleGrid.length;
        int colSize = obstacleGrid[0].length;

        int[][] pathsMatrix = new int[rowSize][colSize];
        
        pathsMatrix[0][0] = 1;

        for (int row = 1; row < rowSize; row++) {
            if (obstacleGrid[row][0] != 1) {
                pathsMatrix[row][0] = pathsMatrix[row - 1][0];
            }
        }

        for (int col = 1; col < colSize; col++) {
            if (obstacleGrid[0][col] != 1) {
                pathsMatrix[0][col] = pathsMatrix[0][col - 1];
            }
        }

        for (int row = 1; row < rowSize; row++) {
            for (int col = 1; col < colSize; col++) {
    
                if (obstacleGrid[row][col] != 1) {
                    pathsMatrix[row][col] = pathsMatrix[row - 1][col] + pathsMatrix[row][col - 1];
                }
            }
        }

        return pathsMatrix[rowSize - 1][colSize - 1];

    }
}
// @lc code=end

